feat: support cloning over SSH via private key auth

johnstcn opened the pull request:
Relates to https://github.com/coder/envbuilder/issues/31

Currently running into issues with testing end-to-end flow.
However, I've tested with a private git repo:

```
$ docker run -it --rm -e GIT_URL="ssh://git@github.com/johnstcn/test-private-devcontainer" -e INIT_SCRIPT=bash -e GIT_SSH_PRIVATE_KEY_PATH=/.ssh/id_ed25519 -v /Users/cian/.ssh/id_ed25519:/.ssh/id_ed25519 envbuilder:latest
envbuilder - Build development environments from repositories in a container
ssh keyscan: github.com ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEmKSENjQEezOmxkZMy7opKgwFB9nkt5YRrYMjNuG5N87uRgg6CLrbo5wAdT/y6v0mKV0U2w0WZ2YB/++Tpockg=
#1: 📦 Cloning ssh://git@github.com/johnstcn/test-private-devcontainer to /workspaces/test-private-devcontainer...
#1: Enumerating objects: 6, done.
#1: Counting objects:  16% (1/6)
#1: Counting objects:  33% (2/6)
#1: Counting objects:  50% (3/6)
#1: Counting objects:  66% (4/6)
#1: Counting obj
#1: ects:  83% (5/6)
#1: Counting objects: 100% (6/6)
#1: Counting objects: 100% (6/6), done.
#1: Compressing objects:  16% (1/6)
#1: Compressing objects:  33% (2/6)
#1: Compressing objects:  50% (3/6)
#1: Compressing objects:  66% (4/6)
#1: Compressing objects:  83% (5/6)
#1: Compressing objects: 100% (6/6)
#1: Compressing objects: 100% (6/6), done.
#1: Total 6 (delta 0), reused 6 (delta 0), pack-reused 0
#1: 📦 Cloned repository! [1.434724587s]
```

One potential issue for which I don't have a solution is that passing an SSH key in as a volume will result in the file being available in the resulting workspace. This may not be desirable; an alternative would be to pass in the SSH key as a base64-encoded file (similar to known_hosts) and clean it up after the build completes.

After this is merged, I'd like to do a follow-up PR to refactor some of the envbuilder setup code.

mafredri commented on envbuilder.go:
```diff
@@ -156,6 +157,34 @@ func Run(ctx context.Context, options Options) error {
+		if err := os.WriteFile(knownHostsPath, knownHostsContent, 0644); err != nil {
+			return fmt.Errorf("write known_hosts file: %w", err)
+		}
+		_ = os.Setenv("SSH_KNOWN_HOSTS", knownHostsPath)
```
Can we add a comment that this is for use by `go-git`?

mafredri commented on envbuilder.go:
```diff
@@ -156,6 +157,34 @@ func Run(ctx context.Context, options Options) error {
+		return fmt.Errorf("invalid git URL: %w", err)
+	}
+	// If we're cloning over SSH, we need a known_hosts file.
+	if gitURLParsed.Scheme == "ssh" {
```
I believe it's possible to clone over SSH even without `ssh://`. Just `git@github.com:org/repo` should suffice. I wonder if this will catch that case?

johnstcn replied:
It will actually refuse to accept that URL, as git URLs are not all valid URLs. 
https://go.dev/play/p/sXvpyxwNLIq

To handle these sorts of URLs, I suppose we could treat all urls without schema prefixes as `ssh://`. WDYT?

BrunoQuaresma replied:
I think it is valid since they don't follow any technical convention format.

mafredri commented on envbuilder.go:
```diff
@@ -156,6 +157,34 @@ func Run(ctx context.Context, options Options) error {
+			}
+		}
+		knownHostsPath := filepath.Join(MagicDir, "known_hosts")
+		if err := os.WriteFile(knownHostsPath, knownHostsContent, 0644); err != nil {
```
Why do we need a file if we set the env?

johnstcn replied:
The default known hosts callback will read the file. I could possibly create a custom known hosts callback that simply iterates over the content of the known_hosts file. WDYT?

mafredri replied:
If it's fairly trivial, go for it 👍🏻. If not, let's at least properly document why we do it.

johnstcn replied:
Even more trivial is simply not doing it!

mafredri commented on git.go:
```diff
@@ -113,3 +122,55 @@ func CloneRepo(ctx context.Context, opts CloneRepoOptions) (bool, error) {
+		return nil, fmt.Errorf("open private key file: %w", err)
+	}
+	var buf bytes.Buffer
+	if _, err := io.Copy(&buf, f); err != nil {
```
Suggestion: `io.ReadAll` would be simpler here.

mafredri commented on git.go:
```diff
@@ -113,3 +122,55 @@ func CloneRepo(ctx context.Context, opts CloneRepoOptions) (bool, error) {
+		if netErr, ok := err.(net.Error); ok {
+			return nil, fmt.Errorf("keyscan %s: %w", dialAddr, netErr)
+		}
+		// If it's not a net.Error then we will assume we were successful.
```
Should we perhaps check if buf is empty?

mafredri commented on git.go:
```diff
@@ -113,3 +122,55 @@ func CloneRepo(ctx context.Context, opts CloneRepoOptions) (bool, error) {
+
+func hostPort(u *url.URL) string {
+	p := 22 // assume default SSH port
+	if _p, err := strconv.Atoi(u.Port()); err == nil {
```
Nit: Prefer `pp`, underscores are unusual in variable names.

BrunoQuaresma replied:
You are using too much TS Cian 😆

mafredri commented on git_test.go:
```diff
@@ -159,6 +164,101 @@ func TestCloneRepo(t *testing.T) {
 }
 
+func TestCloneRepoSSH(t *testing.T) {
+
```
```suggestion
```
Maybe we should introduce [gofumpt](https://github.com/mvdan/gofumpt) in this repo. 🤔

mafredri reviewed (commented):
> an alternative would be to pass in the SSH key as a base64-encoded file (similar to known_hosts) and clean it up after the build completes.

I think this would be a solid first step at least. We can always iterate/add support for other use-cases later.

BrunoQuaresma commented on envbuilder.go:
```diff
@@ -156,6 +157,34 @@ func Run(ctx context.Context, options Options) error {
+	// If we're cloning over SSH, we need a known_hosts file.
+	if gitURLParsed.Scheme == "ssh" {
+		var knownHostsContent []byte
+		if options.GitSSHKnownHostsBase64 != "" {
```
Is it sensitive data? Asking because I see it is encoded on base 64 and I'm wondering why.

johnstcn replied:
It's easier to pass data as an environment variable that way; it also follows the existing convention with the docker config JSON.

BrunoQuaresma commented on envbuilder.go:
```diff
@@ -156,6 +157,34 @@ func Run(ctx context.Context, options Options) error {
+		var knownHostsContent []byte
+		if options.GitSSHKnownHostsBase64 != "" {
+			if kh, err := base64.StdEncoding.DecodeString(options.GitSSHKnownHostsBase64); err != nil {
+				return fmt.Errorf("invalid known_hosts content: %w", err)
```
Would make sense to make it more explicit like `invalid Base64 content`? As a user, I would think my `know_hosts` is using some wrong URL or maybe there is a typo error in the config.

johnstcn replied:
Makes sense!

BrunoQuaresma commented on testutil/gittest/gittest.go:
```diff
@@ -97,6 +105,97 @@ func NewServer(fs billy.Filesystem) http.Handler {
+	}
+
+	done := make(chan struct{}, 1)
+	go func() {
```
I guess this is some go routine stuff right?

johnstcn replied:
Yeah we don't want `ssh.Serve()` to block execution but we want to wait for it to finish before continuing.

A common pattern here is to use use the fact that recieving on a closed channel will always immediately return the zero value for the channel. So you'l often see stuff like this:

```
done := make(chan struct{})
go func() {
  someBlockingFunction()
  close(done)
}()

<-done // waits for someBlockingFunction to finish
```

BrunoQuaresma reviewed (commented):
Left a few comments but mostly are for my own learning 🙏

johnstcn commented on git.go:
```diff
@@ -113,3 +122,55 @@ func CloneRepo(ctx context.Context, opts CloneRepoOptions) (bool, error) {
+
+// GenerateKnownHosts dials the server located at gitURL and fetches the SSH
+// public keys returned in a format accepted by known_hosts.
+func GenerateKnownHosts(log LoggerFunc, gitURL *url.URL) ([]byte, error) {
```
Unfortunately I don't really have a good answer for that, apart from that's what most functions that operate on URLs want, and that's what `url.Parse()` returns. https://pkg.go.dev/net/url#Parse

dannykopping reviewed (approved):
LGTM! Mostly nits

mafredri commented on envbuilder.go:
```diff
@@ -156,6 +157,36 @@ func Run(ctx context.Context, options Options) error {
+			if options.GitSSHKnownHostsBase64 != "" {
+				if kh, err := base64.StdEncoding.DecodeString(options.GitSSHKnownHostsBase64); err != nil {
+					return fmt.Errorf("invalid known_hosts content: %w", err)
+				} else {
+					knownHostsContent = kh
+				}
```
Normally I'd agree and prefer to avoiding `else`, but since `kh` is assigned in the `if`-statement, it needs to be in the `else` or then the assignment moved out of `if`. Both are alright approaches IMO.

johnstcn replied:
I'm going to simplify this logic a bit.

johnstcn commented on git_test.go:
```diff
@@ -159,6 +164,136 @@ func TestCloneRepo(t *testing.T) {
+func TestCloneRepoSSH(t *testing.T) {
+	t.Parallel()
+
+	t.Run("AuthSuccess", func(t *testing.T) {
```
Agreed in principle; but I'd prefer to do this when I have more test cases from which to abstract.

johnstcn commented:
I've simplifed the logic by removing the need to parse the git URL at all and removing the known_hosts business; if folks need to validate host keys they can simply mount in a known_hosts file and set SSH_KNOWN_HOSTS.
I've also extracted the logic for setting up repo auth into a separate function and added dedicated tests for this.

johnstcn commented:
Bonus points: SSH_AUTH_SOCK also works!

```
docker run -it --rm -e GIT_URL="ssh://git@github.com/johnstcn/test-private-devcontainer" -e SSH_AUTH_SOCK=/tmp/ssh-Vsi58Zsh72/agent.1567 -v /tmp/ssh-Vsi58Zsh72/agent.1567:/tmp/ssh-Vsi58Zsh72/agent.1567 -e INIT_SCRIPT=bash envbuilder:latest
envbuilder - Build development environments from repositories in a container
#1: 📦 Cloning ssh://git@github.com/johnstcn/test-private-devcontainer to /workspaces/test-private-devcontainer...
#1: 🔑 Using SSH authentication!
#1: 🔑 No SSH key found, falling back to agent!
#1: 🔓 SSH_KNOWN_HOSTS not set, accepting all host keys!
#1: 🔑 Got host key: github.com,20.26.156.215 ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEmKSENjQEezOmxkZMy7opKgwFB9nkt5YRrYMjNuG5N87uRgg6CLrbo5wAdT/y6v0mKV0U2w0WZ2YB/++Tpockg=
```

dannykopping commented on git_test.go:
```diff
@@ -259,39 +260,115 @@ func TestCloneRepoSSH(t *testing.T) {
+		require.Equal(t, actualSigner, pk.Signer)
+	})
+
+	t.Run("SSH/NoAuthMethods", func(t *testing.T) {
```
Can you add a case for an invalid scheme?

dannykopping commented on git_test.go:
```diff
@@ -259,39 +260,115 @@ func TestCloneRepoSSH(t *testing.T) {
-func TestParseGitURL(t *testing.T) {
-	t.Parallel()
+// nolint:paralleltest // t.Setenv for SSH_AUTH_SOCK
+func TestSetupRepoAuth(t *testing.T) {
```
Super!!

dannykopping reviewed (approved):
Slick! Love the changes, last comment from me

mafredri commented on git.go:
```diff
@@ -113,3 +123,131 @@ func CloneRepo(ctx context.Context, opts CloneRepoOptions) (bool, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("open private key file: %w", err)
+	}
```
```suggestion
	}
	defer f.Close()
```

mafredri commented on git.go:
```diff
@@ -113,3 +123,131 @@ func CloneRepo(ctx context.Context, opts CloneRepoOptions) (bool, error) {
+			options.Logger(codersdk.LogLevelError, "#1: ❌ Failed to connect to SSH agent: %s", err.Error())
+			return nil // nothing else we can do
+		}
+		if os.Getenv("SSH_KNOWN_HOSTS") == "" {
```
Suggestion: If this was moved higher up, we could avoid duplicating it.

mafredri commented on git_test.go:
```diff
@@ -159,6 +165,212 @@ func TestCloneRepo(t *testing.T) {
+		require.Nil(t, auth)
+	})
+
+	t.Run("HTTP/NoAuth", func(t *testing.T) {
```
Suggestion: could avoid prefix if these were nested in an otherwise empty `t.Run`. Feel free to disregard if you don't want to do the change. 😄

johnstcn replied:
I think one level of nesting is as deep as we need here in the test :-)

mafredri commented on git.go:
```diff
@@ -113,3 +123,131 @@ func CloneRepo(ctx context.Context, opts CloneRepoOptions) (bool, error) {
+	// GIT_USERNAME if set.
+	authUser := options.GitUsername
+	if authUser == "" {
+		authUser = "git"
```
Suggestion: consider updating `options.GitUsername` if empty instead, so that we don't need to duplicate this check and can use the option directly.

mafredri reviewed (approved):
Looks great!
